### PR TITLE
fix: carry hold outlines through tiling, and add a trainable seg config

### PR DESCRIPTION
**Pull request by marcodejongh**
## Summary

`ml/holds/data/tile_coco.py` clipped every box to its tile and dropped `segmentation` entirely. Nothing failed — a tiled segmentation config trained its mask head on empty polygons and never learned an outline, which reads as a model problem rather than a data one. Polygons are now clipped with Sutherland–Hodgman against the tile rectangle: exact for a rectangular window, and no new dependency for one crop.

Verified on the corpus this unblocks: **96.6% of annotations keep a polygon through tiling** (identical to untiled) and **no polygon lands outside its tile**.

Also adds `seg-nano-untiled-1024` as the resolution control next to the existing tiled config. Its notes record the trap: seg-nano is `patch_size 12` with `num_windows 1`, so a 768 input is global attention over 4096 tokens — 56× nano's cost at the same resolution — and it OOM'd twice on an M5 Max at 112 GiB before the resolution came down to 384.

### Why this matters

With tiling fixed, `seg-nano-tiled-1024` trains and produces its own hold polygons, which is what `spray_wall_holds.outline` wants (NULL there falls back to drawing the hold as a plain circle).

Scored on the held-out spray-wall eval split, both int8 ONNX, threshold chosen on the tune split:

| | nano box detector (ships today) | seg-nano tiled |
| --- | --- | --- |
| Box F1 | 0.659 | 0.647 |
| Fixes per hold | 0.660 | 0.673 |
| Outline source | classical in-box guess | model mask |
| Median outline fill | 0.477 | 0.698 |
| Detections given an outline | n/a | 99.7% |
| int8 artifact | 31 MB | 31 MB |
| CPU latency / photo | 0.166 s | 0.316 s (4 tiles) |

Detection is unchanged within this 10-photo split's error bars (the bootstrap puts ~±0.05 on a difference of this kind); the outlines are not.

Masks for training were distilled: Roboflow's **human** boxes kept exactly, with only the outline inside each box generated by Crux (MIT) through `forward_with_given_boxes`. No hold is invented or missed, so the usual pseudo-labelling failure — missed holds becoming background — cannot occur. Provenance is written into the annotation file.

`train.py`'s cache-key collision, which I also hit, turned out to be already fixed on main by `tiled_dataset_dir()` — that change is dropped from this branch.

## Test plan

1. CI green.

## Risk

Risk: 2/5 — offline training tooling under `ml/holds/`; nothing in this PR ships to a device or runs in the app.

## Release Notes

- [x] No release note needed (internal / technical change)

### CODE REVIEW

**Comment by marcodejongh** [comment]
Addressed the publisher-schema nit in e77643344: publication validation now builds its own shallow schema copy and a deduplicated required-field list. The source schema was already reloaded on every call, but this also remains safe if it is cached later.

The other two findings refer to behavior already covered by the pushed change:

- `ml/holds/test_tile_coco.py::test_rle_masks_fail_explicitly_instead_of_losing_training_labels` parametrizes both list-count and compressed-string RLE dictionaries and asserts the explicit `ValueError`.
- `outputs.masks.required` includes both `activation` and `layout`. Replacing them with the review's `activtion` or `laoyut` examples fails validation; both examples were checked directly. Extra additive fields stay allowed for schema-v1 compatibility, matching the existing boxes/logits output contract. The mask decode order remains pinned by a constant and its regression test.

The focused Python suite passes 78 tests total after the update. This is the total suite count, not 78 newly added tests. Full `vp check` and the 53-task `vp run typecheck` passed before this Python-only schema-copy refinement.

File: ml/holds/data/tile_coco.py [hunk 1]
```diff
@@ -72,6 +156,9 @@ def tile_split(source: Path, target: Path, grid: TileGrid, min_visible: float, l
                         {
                             "bbox": [clipped_x0 - x0, clipped_y0 - y0, clipped_w, clipped_h],
                             "area": clipped_w * clipped_h,
+                            "segmentation": clip_polygon_to_tile(
+                                annotation.get("segmentation"), scale, x0, y0, x1, y1
+                            ),
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply ml/holds/data/tile_coco.py:165]: Fixed in 9a2554bae. RF-DETR seg-* training now requires a usable positive-area polygon for every annotated hold before tiling or model construction. The same validation checks prepared tiles, matching cached datasets and the actual capped training subset, while box detectors and classical postprocessing retain box-only support.

96 lightweight Python tests pass, including 18 new regressions for box-only/mixed/empty labels, validation/test splits, stale caches, actual polygon tiling and reuse, main() refusing model construction, and an all-negative capped subset. Full vp check passed with zero errors; full vp run typecheck passed all 53 tasks; independent paired review and commit hooks passed. No model was trained or published. README and PR body now explicitly state that the default Way Up box-only corpus cannot train mask models and historical 96.6% coverage does not satisfy the new per-hold requirement.

**Comment by marcodejongh** [comment]
Addressed the fresh review in 9a4153d7d. The fractional collinear ring `[0.01,0.03,0.02,0.05,0.03,0.07]` reproduced the concern: its floating-point shoelace sum was `1.0842021724855044e-19`. After coordinate rounding, the area check now uses integer hundredths, which rejects that ring exactly and preserves the smallest valid grid triangle. The new regressions cover both outcomes. Source-label guidance now says to correct source annotations before rebuilding affected caches.

Checked `docs/spray-recognition-rollout.md`: its Current architecture section already documents the homelab Node worker, writable primary, private R2 reads and shared outline decoder; its deployment/TLS/exposure gates remain explicitly unsatisfied. No rollout-document architecture change was necessary.

All 98 lightweight Python tests, full vp check (zero errors), full vp run typecheck (53 tasks), repository hook and independent paired review passed. No new model training, corpus evaluation, publication or production change was performed.

**Comment by marcodejongh** [comment]
Addressed the retry and ConfigInfo notes in `96fbbcb8f`. An unstamped tiled cache is revalidated first, so a retry after failed mask validation reports the original useful mask error. If the masks are valid but the stamp is missing, preparation reports missing provenance and still refuses reuse. Failed files remain available for inspection; no cache deletion or success-stamp fabrication was added. The regression now calls preparation twice and proves only one tiler invocation, retained files and no stamp; a separate valid-mask/unstamped-cache case also refuses reuse. `long_side` is now a required ConfigInfo field.

The validation paths are mutually exclusive: preparation checks the source plus either the cached dataset or the freshly generated dataset. Capping adds one check of the actual selected subset, which can lose all labelled targets even when its source is valid. Those checks remain because their regressions cover distinct failures; there is no normal source+cached+freshly-generated triple pass.

All 99 light Python tests passed, along with full lint/format, all 53 typecheck tasks and the commit hook. Independent paired review approved the four-file change. No training run, model download, production change or new quality claim was made.

**Comment by marcodejongh** [comment]
Separated malformed COCO image metadata from mask-label failures. Orphaned image references and missing width/height now identify the file, annotation and structural problem; they no longer tell the operator to replace valid polygon labels with a different corpus. Three regressions fail against the previous diagnostic and pass with the fix. All 102 lightweight Python tests, full lint/format, all 53 typecheck tasks and the commit hook passed, with independent paired review.

The geometry check is necessary: three or more coordinate pairs do not establish a usable mask. Collinear vertices have zero area, and a polygon wholly outside the image clips to nothing. `test_missing_outside_and_degenerate_polygons_produce_no_ring`, `test_fractional_collinear_points_do_not_survive_floating_point_area_noise`, and the training entry-point tests cover those distinctions. Reusing the actual clipping/rounded-coordinate rules keeps validation consistent with what the trainer receives, so the proposed structural-only shortcut was not applied.

No model training, dataset download or new quality claim was made.

**Comment by marcodejongh** [comment]
Checked the three optional notes against the current entry points:

- Direct tiler callers intentionally receive `ValueError("Expected COCO polygon lists; RLE segmentation is not supported")`; this is an explicit unsupported-format failure, not a silent conversion. The lightweight-preprocessing section already requires converting RLE to polygon annotations. The training entry point adds file/annotation context before invoking the tiler. The direct API retains its exception contract.
- The mask preflight requires training annotations and validates every annotated hold in each supplied validation/test annotation file. It does not certify that a corpus contains every split required by a chosen training/evaluation workflow. Missing optional files are therefore skipped; malformed files that are present fail. The separate evaluation workflow remains responsible for its requested evaluation split, and no evaluation result is inferred from the training guard.
- The publisher does not accept an existing manifest as its input: `main()` always calls `build_manifest()` from the chosen config before validation, including `--force`. That fresh manifest always carries `inference`. `test_force_allows_overwrite`, `test_force_replaces_only_the_manifest`, and the legacy-weight tests cover republishing; force still cannot replace different immutable weight bytes. The shared reader schema remains compatible with older published manifests that lack the additive field.

These clarifications require no change to mask acceptance or publishing behavior. The current 102-test Python suite passes.
